### DISCUSSION TITLE
store/repository.goのNew関数内でのPing失敗時のリターンのdb.Closeについて(コード47行目)

### DISCUSSION THREAD

**Discussion by yamagit01** [budougumi0617/go_todo_app#81] [Ideas]
とてもわかりやすい書籍を執筆いただきありがとうございます。
勉強させていただいています。

指摘についてですが、
タイトルの箇所について、後々DBをクローズできるようにfuncでリターンしていますが、
このリターンはNewMux関数内でcleanupとして受け、そのままmainのrun関数に返されています。
run関数内ではエラー判定の後にdefer文でcleanupの実行を指定していますので、エラー時はクローズ処理が実行されない認識です。
ただ、単純にdeferをエラー判定の前に置くと、sql.Openのエラー時はcleanupにnilが返ってくるためpanicになるので、
sql.Openのエラー時にfunc(){}を返せばとりあえずは上手くいくのではないかと思っています。
（defer文はエラー判定の後に記述するのが定石な気がするのでdeferが前にあると不自然な気はしますが。。。）
通常どのように記述するのが良いのかご教授いただければ幸いです。

後、付随した質問なのですが、今回のsql.Openのエラー時のreturnのように返り値がfunc()の際に
nilを返すことは開発する際に結構あるのでしょうか。後続で実行されてpanic等ならないようfunc(){}のように
空の値を入れておくほうが良いのかなと思っているのですが。
以上、よろしくおねがいします。

**Comment by budougumi0617** [comment]
> sql.Openのエラー時にfunc(){}を返せばとりあえずは上手くいくのではないか

こちらそのとおりだと思います！受け手が「 `cleanup` 用の戻り値がnilの可能性を考慮する」というのはわずらわしいことなので、ご指摘いただいたとおりに実装するのがよいですね。

> defer文はエラー判定の後に記述するのが定石な気がするのでdeferが前にあると不自然な気はしますが。。。

たしかに不自然さではあるものの、記述的には一番シンプルかなと思いました。
何も知らない人が読むと「これは実装ミスだな」と思って修正してしまう可能性もあるので、補足コメントをしてエラー判定の前に書くのが良いかなと思いました。
https://github.com/budougumi0617/go_todo_app/blob/v1.0.12/main.go#L31-L36

**Reply by yamagit01** [reply]
ご回答ありがとうございます。
確かに提示していただいたような補足コメントがあれば問題ないですね。
勉強になりました。

**Comment by budougumi0617** [comment]
> 今回のsql.Openのエラー時のreturnのように返り値がfunc()の際に
> nilを返すことは開発する際に結構あるのでしょうか。後続で実行されてpanic等ならないようfunc(){}のように
> 空の値を入れておくほうが良いのかなと思っているのですが。

私もそのとおりだと思います。なので nilを返さず空関数を返したほうがよいと思います。
実務でも似たような実装をしているところがあるのですが、書籍のとおりあまり意識せずに実装していたので見直しておこうと思います。

もろもろ正誤表やコードに反映させていただきました。ありがとうございます！
https://github.com/budougumi0617/go_todo_app/releases/tag/v1.0.12

**Reply by yamagit01** [reply]
やはり、安全に倒すほうが良いですよね。
反映していただきありがとうございます。
